.github/workflows/ci.yml

```diff
@@ -34,6 +34,98 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      - name: Setting build env
+        run: |
+          echo "RUST_OS_NAME=${{ runner.os }}" >> $GITHUB_ENV
+        shell: bash
+
+      # We install gnu-tar because BSD tar is buggy on Github's macos machines.
+      # See <https://github.com/actions/cache/issues/403>.
+      - name: Install GNU tar
+        if: runner.os == 'macOS'
+        run: |
+          brew install gnu-tar
+          echo "/usr/local/opt/gnu-tar/libexec/gnubin" >> $GITHUB_PATH
+
+      # Cache the global cargo directory, but NOT the local `target` directory which
+      # we cannot reuse anyway when the nightly changes (and it grows quite large
+      # over time).
+      - name: Add cache for cargo
+        uses: actions/cache@v2
+        with:
+          path: |
+            # to avoid saving rustup proxy, save xargo and RTIM in separate dir.
+            ~/.cargo/ci
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: ${{ runner.os }}-cargo
+
+      # We use the "stable" toolchain for better caching, it is just used to build `rustup-toolchain-install-master`.
+      # But we also need to take into account that the build cache might have a different, outdated default.
+      - name: Install stable Rust
+        run: |
+          rustup toolchain install stable
+          rustup default stable
+
+      - name: Install RTIM
+        run: |
+          cargo install --root ~/.cargo/ci rustup-toolchain-install-master
+          echo "$HOME/.cargo/ci/bin" >> $GITHUB_PATH
+        shell: bash
+
+      - name: Install xargo
+        run: |
+          cargo install --root ~/.cargo/ci xargo
+        shell: bash
+
+      - name: Install "master" toolchain
+        run: |
+          if [[ ${{ github.event_name }} == 'schedule' ]]; then
+            RUSTC_HASH=$(git ls-remote https://github.com/rust-lang/rust.git master | awk '{print $1}')
+          else
+            RUSTC_HASH=$(< rust-version)
+          fi
+          rustup-toolchain-install-master \
+            -f \
+            -n master "$RUSTC_HASH" \
+            -c rust-src \
+            -c rustc-dev \
+            -c llvm-tools \
+            --host ${{ matrix.host_target }}
+          rustup default master
+        shell: bash
+
+      - name: Inspect Rust version
+        run: |
+          rustup show
+          rustc -Vv
+          cargo -V
+
+      - name: Test
+        run: bash ./ci.sh
+
+      # Don't cache "master" toolchain, it's a waste.
```

Comment:
The step does not seem to have anything to do with the comment?

Reply:
yeah, it's obsolete. By default, GH Actions restores .rustup and .cargo from their images.
There is not much needed to cache stable toolchain in .rustup.

Reply:
The point is to delete the master toolchain before creating the cache. Your script does not seem to do that?

Reply:
If we don't save .rustup, there is no need to do that.

Reply:
That is true, but... doesn't that mean we download the stable toolchain each time, instead of using the cached version?

Reply:
gh actions updates rustup stable toolchain (6 weeks or so) each time they update their OS images.
also, I don't think there is much reason in caching .rustup dir. The downloading time and extraction time
from CI caches is equivalent to installation time from rustup.

Reply:
>  The downloading time and extraction time
from CI caches is equivalent to installation time from rustup.

The cache is presumably downloaded from inside the GHA infrastructure, it does not have to be downloaded from the official rust-lang servers.

Do you know if GHA patches rustup to avoid hitting the rust-lang servers each time?

> gh actions updates rustup stable toolchain (6 weeks or so) each time they update their OS images.

So the stable is already in the OS image...? Is there documentation on that?

Reply:
The documentation is <https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-software>. You could click each runner OS to see the preinstalled software list.

> if GHA patches rustup to avoid hitting the rust-lang servers each time

They download and use rustup from rustup.rs. So nope, they doesn't patch rustup.

> The cache is presumably downloaded from inside the GHA infrastructure

Yeah, I could measure the cache time after caching .rustup if you want.
Another problem when caching .rustup on GHA is that we have to check the rustc build hash to know
when to refresh the .rustup cache if stable updated. The check is not that complicated but I think
you are not a fan of it: <https://github.com/rust-lang/rustup/blob/243b453934929ea19f4172f626018531bab45ddd/ci/actions-templates/linux-builds-template.yaml#L95-L105>

Reply:
> Another problem when caching .rustup on GHA is that we have to check the rustc build hash to know
when to refresh the .rustup cache if stable updated

But why that? Travis and AppVeyor have been cashing the rustup dir and they did not need anything like that.

It is totally okay if the cache contains an outdated stable. We'll just update it when CI runs then. I don't see any reason why this needs to be reflected in the cache key.

Reply:
You figured it out: <https://github.com/rust-lang/miri/pull/1571#discussion_r502787658>.

> It is totally okay if the cache contains an outdated stable.

Yeah, GHA already does that for us.